pkg/manager/member/tikv_upgrader.go

```diff
@@ -270,29 +273,140 @@ func (u *tikvUpgrader) modifyVolumesBeforeUpgrade(tc *v1alpha1.TidbCluster, upgr
 	return true, nil
 }
 
+func (u *tikvUpgrader) endEvictLeaderAfterUpgrade(tc *v1alpha1.TidbCluster, pod *corev1.Pod) (bool /*done*/, error) {
+	logPrefix := fmt.Sprintf("endEvictLeaderAfterUpgrade: for tikv pod %s/%s", pod.Namespace, pod.Name)
+
+	store, err := TiKVStoreFromStatus(tc, pod.Name)
+	if err != nil {
+		return false, err
+	}
+	storeID, err := strconv.ParseUint(store.ID, 10, 64)
+	if err != nil {
+		return false, err
+	}
+
+	// evict leader
+	err = u.endEvictLeader(tc, storeID, pod)
+	if err != nil {
+		return false, err
+	}
+
+	// wait for leaders to transfer back or timeout
+	// refer to https://github.com/pingcap/tiup/pull/2051
+
+	isLeaderTransferBackOrTimeout := func() bool {
+		evictLeaderEndTimeStr, exist := pod.Annotations[annoKeyEvictLeaderEndTime]
+		if !exist {
+			klog.Errorf("%s: miss annotation %q, so skip waiting leaders for transfer back", logPrefix, annoKeyEvictLeaderEndTime)
+			return true
+		}
+
+		evictLeaderEndTime, err := time.Parse(time.RFC3339, evictLeaderEndTimeStr)
+		if err != nil {
+			klog.Errorf("%s: parse annotation %q to time failed, so skip waiting leaders for transfer back", logPrefix, annoKeyEvictLeaderEndTime)
+			return true
+		}
+
+		leaderCountBefore := int(*store.LeaderCountBeforeUpgrade)
+		if leaderCountBefore < 200 {
+			klog.Infof("%s: leader count is %d and less than 200, so skip waiting leaders for transfer back", logPrefix, leaderCountBefore)
+			return true
+		}
```

Comment:
This can be moved to the head of `isLeaderTransferBackOrTimeout`.

The parse of `evictLeaderEndTime` is before this check but the usage of `evictLeaderEndTime` is after this check.